feat: sprint7 step2

dmitry-razumov opened the pull request:
PR for review

dmitry-razumov committed:
feat: sprint7 step2

ArslanMustafin commented on src/components/burger-ingredients/burger-ingredients.jsx:
```diff
@@ -0,0 +1,66 @@
+  function ShowTab() {
+    const [current, setCurrent] = React.useState('Булки')
+    return (
+      <div style={{ display: 'flex' }}>
```
**Надо исправить:**

Не стоит писать стили инлайново, в будущем это может затруднить поиск ошибок, и перебить стили заданных классов. Лучше дать какой-то класс, и там указать уже нужные вам стили.

ArslanMustafin commented on src/components/app-header/app-header.jsx:
```diff
@@ -0,0 +1,41 @@
+import { BurgerIcon } from '@ya.praktikum/react-developer-burger-ui-components/dist/ui/icons';
+import { ListIcon } from '@ya.praktikum/react-developer-burger-ui-components/dist/ui/icons';
+import { ProfileIcon } from '@ya.praktikum/react-developer-burger-ui-components/dist/ui/icons';
+import { isActive } from '../app/app';
```
**Можно лучше:**

Лучше перенести переменную сюда, т.к. именно она отвечает за вид ссылок и используется только в этом файле.

ArslanMustafin commented on src/components/app/app.jsx:
```diff
@@ -1,16 +1,32 @@
+export const isActive = true;
 
 function App() {
+  const [data, setData] = React.useState({
```
**Можно лучше:**

Тут и далее, лучше импортировать функции отдельно, а не использовать их через React.

```js
import { useState, useEffect } from "react"; 
```

ArslanMustafin commented on src/utils/data.js:
```diff
@@ -1,3 +1,6 @@
+export const baseUrl = "https://norma.nomoreparties.space/api/ingredients";
+export const orderNumber = Number('034536');
```
**Надо исправить:**

Константы принято записывать в `UPPER_CASE`

ArslanMustafin replied:
**Можно лучше:**

Лучше написать так:

```js
export const orderNumber = 034536;
```

ArslanMustafin commented on src/utils/data.js:
```diff
@@ -1,3 +1,6 @@
+export const baseUrl = "https://norma.nomoreparties.space/api/ingredients";
+export const orderNumber = Number('034536');
+
 export const data = [
```
**Можно лучше:**

думаю это можно уже удалить

ArslanMustafin commented on src/components/app/app.jsx:
```diff
@@ -1,16 +1,32 @@
-      	Измените src/components/app/app.jsx и сохраните для обновления.
-      </pre>
+      <AppHeader />
+      {!data.isLoading && !data.isError && <Main ingredients={data.ingredientsData} />}
+      {data.isLoading && !data.isError && <p className={`text text_type_main-large ${styles.loading}`}>Данные загружаются</p>}
+      {data.isError && <p className={`text text_type_main-large ${styles.loading}`}>{`Ошибка сервера: ${data.errorType}`}</p>}
```
**Можно лучше:**

Можно написать чуточку лучше.

```js
 React.useEffect((data) => {
    getIngredientsData(data)
      .then(res => {
        setData((prevData) => ({...prevData, ingredientsData: res}))
      })
       .catch(() => {
        setData((prevData) => ({...prevData, isError: res}))
      })
      .finally(() => {
        setData((prevData) => ({...prevData, isLoading: false}))
      });;
  }, [])

  return (
    // ...

    {data.isLoading && <...>} // лоадер
    {data.isError && <...>} // ошибка
    {data.ingredientsData && <...>} // основной контент

)
```

Так вам не придётся делать условия по несколько раз, и это будет оптимальным вариантом.

ArslanMustafin commented on src/components/app/app.jsx:
```diff
@@ -1,16 +1,32 @@
+    errorType: ''
+  })
+
+  React.useEffect((data) => {
+    getIngredientsData(data, setData);
+  }, [])
```
**Надо исправить:**

Не стоит передавать какие-либо функции в функцию запроса. Функция запроса должна делать только запрос (ну и обработать ошибку если она есть) и вернуть Promise. Вот с промисом можно делать всё что угодно)) а то вы сильно ограничиваете переиспользование функции. Возьмите за правило, если вы работаете с запросами, то возвращайте промис из функций, и в дальнейшем работайте с ним.

```js
 React.useEffect((data) => {
    getIngredientsData(data)
      .then(res => {
        setData((prevData) => ({...prevData, ingredientsData: res}))
      })
       .catch(() => {
        setData((prevData) => ({...prevData, isError: res}))
      })
      .finally(() => {
        setData((prevData) => ({...prevData, isLoading: false}))
      });;
  }, [])
```

ArslanMustafin commented on src/components/burger-constructor/burger-constructor.jsx:
```diff
@@ -0,0 +1,101 @@
+import Modal from './../modal/modal';
+
+function BurgerConstructor({ ingredients }) {
+  const product = [
+    {type:'bun', name: 'Булки'},
+    {type: 'sauce', name: 'Соусы'},
+    {type: 'main', name : 'Начинки'}
+  ]
```
**Можно лучше:**

Лучше вынести за пределы компонента. Чтобы при ререндере объект не создавался заново.

ArslanMustafin commented on src/components/burger-constructor/burger-constructor.jsx:
```diff
@@ -0,0 +1,101 @@
+    setModalActive(false);
+  };
+
+  const bunFirst = ingredients.findIndex(item => item.type === product[0].type && item.quantity !== 0);
+  const bunAny = ingredients.findIndex(item => item.type === product[0].type);
+  const bunIndex = bunFirst === -1 ? bunAny : bunFirst;
```
**Надо исправить:**

Булки у нас в любом случае будут одинаковые. Поэтому можно просто написать вот так.

```js
const { bun, ingredients } = useMemo(() => {
  return {
    bun: data.find(item => item.type === 'bun'),
    ingredients: data.filter(item => item.type !== 'bun'),
  };
}, [data]);
```

ArslanMustafin commented on src/components/burger-constructor/burger-constructor.jsx:
```diff
@@ -0,0 +1,101 @@
+  const bunFirst = ingredients.findIndex(item => item.type === product[0].type && item.quantity !== 0);
+  const bunAny = ingredients.findIndex(item => item.type === product[0].type);
+  const bunIndex = bunFirst === -1 ? bunAny : bunFirst;
+  ingredients[bunIndex].quantity = 2;
```
**Можно лучше:**

на первый раз ок, но в будущем, через спринт, вам это нужно будет исправить)) 
Не очень хорошо мутировать объект в реакте таким способом, реакт ведь не будет знать что объект обновился, и соответсвенно не будет перерисовывать элементы.

ArslanMustafin commented on src/components/burger-constructor/burger-constructor.jsx:
```diff
@@ -0,0 +1,101 @@
+  
+  const totalPrice = ingredients.map((item) => item.price * item.quantity).reduce((acc, item) => acc + item, 0);
+
+  function ShowCreateBurger({ item, kind, lock }) {
+    const newtext = (kind === "top") ? [item.name, " (верх)"].join('') :
+      (kind === "bottom") ? [item.name, " (низ)"].join('') : item.name;
+    return (
+      <ConstructorElement
+        type={kind}
+        isLocked={lock}
+        text={newtext}
+        price={item.price}
+        thumbnail={item.image}
+      />
+    )
+  }
```
**Надо исправить:**

Лучше не создавать отдельные функции для рендера элементов, это не совсем правильно.
Намного лучше и оптимальнее будет просто использовать уже созданный компонент.

```jsx
  <ConstructorElement
    type="top"
    text={`${bun.name} (верх)`}
    price={bun.price}
    thumbnail={bun.image}
    isLocked
  />
```

ArslanMustafin commented on src/components/burger-constructor/burger-constructor.jsx:
```diff
@@ -0,0 +1,101 @@
+    )
+  }
+
+  function ShowItem({ item, kind, lock}) {
+    return (
+      <section className={styles.draggable}>
+        {!lock && < DragIcon />}
+        <ShowCreateBurger item={item} kind={kind} lock={lock} />
+      </section>
+    );
+  };
```
**Надо исправить:**

Лучше не создавать отдельные функции для рендера элементов, это не совсем правильно.
Намного лучше и оптимальнее будет просто использовать вашу структуру в чистом виде. 
Тем более она используется только в этом месте.

Но если вы хотите как-то это вынести, то выносите в отдельный компонент (в папку и файл), и не забудьте про propTypes.

ArslanMustafin commented on src/utils/prop-types.js:
```diff
@@ -2,4 +2,18 @@ import PropTypes from "prop-types";
+  __v: PropTypes.number.isRequired
 });
+
+export const dataPropType =  PropTypes.arrayOf(ingredientPropType); 
```
**Надо исправить:**

Необходимо указывать .isRequired как у элемента массива, так и у самого массива в propTypes

```js
export const dataPropType =  PropTypes.arrayOf(ingredientPropType.isRequired); 
```

ArslanMustafin commented on src/components/burger-ingredients/burger-ingredients.jsx:
```diff
@@ -0,0 +1,66 @@
+    const [current, setCurrent] = React.useState('Булки')
+    return (
+      <div style={{ display: 'flex' }}>
+        <Tab value='Булки' active={current === 'Булки'} onClick={setCurrent} >
+          {product[0].name}
+        </Tab>
+        <Tab value='Соусы' active={current === 'Соусы'} onClick={setCurrent}>
+          {product[1].name}
+        </Tab>
+        <Tab value='Начинки' active={current === 'Начинки'} onClick={setCurrent}>
+          {product[2].name}
+        </Tab>
```
**Можно лучше:**

а почему бы тут не использовать вывод по циклу?

```js
{products.map((product) => (
  <Tab value='Булки' active={current === product.type} onClick={setCurrent} >
    {product.name}
  </Tab>
))}
```

ArslanMustafin commented on src/components/burger-ingredients/burger-ingredients.jsx:
```diff
@@ -0,0 +1,66 @@
+
+
+function BurgerIngredients({ingredients}) {
+  const product = [
```
**Надо исправить:**

У вас тут массив, поэтому название должно быть в мн.числе `products`

ArslanMustafin commented on src/components/burger-ingredients/burger-ingredients.jsx:
```diff
@@ -0,0 +1,66 @@
+    )
+  }
+
+  function ShowItem({ data, product}) {
+    return (
+      <>
+        <h2 className={styles.headline}>{product.name}</h2>
+        <div className={styles.grid}>  {
+          data.map((item) => ((item.type === product.type) &&
+            <IngridientItem key={item._id} item={item}>
+              <img src={item.image} alt={item.name} />
+            </IngridientItem>
+          ))
+        }
+        </div>
+      </>
+    );
+  }
```
**Надо исправить:**

это нужно вынести в отдельный компонент (папку и файл), не забудьте и про propTypes.

ArslanMustafin commented on src/components/ingredient-details/ingredient-details.jsx:
```diff
@@ -0,0 +1,35 @@
+import { ingredientPropType } from './../../utils/prop-types';
+import styles from './ingredient-details.module.css';
+
+function IngredientDetails(data) {
```
**Надо исправить:**

Лучше деструктурировать пропсы, а то `data.data`, выглядит не очень.

ArslanMustafin commented on src/components/main/main.jsx:
```diff
@@ -0,0 +1,26 @@
+import BurgerConstructor from './../burger-constructor/burger-constructor';
+
+const Main = (props) => {
+  const localIngredients = props.ingredients.slice();
+  localIngredients.forEach((item) => item.quantity = 0);
```
**Надо исправить:**

если вы хотите изменить элементы массива, то нужно использовать `map` или `reduce`, ни в коем случае не используте для этого for, это мутирует исходный массив.

```js
const localInggredients = props.ingredients.map((item) => ({...item, quantity: 0}))
```

Подробнее про [map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map) и [reduce](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/reduce).

ArslanMustafin commented on src/components/modal/modal.jsx:
```diff
@@ -0,0 +1,47 @@
+const modalRoot = document.getElementById('react-modals');
+
+const Modal = ({ title, children, closeModal }) => {
+  const closeByEsc = (event) => {
+    if (event.key === "Escape") {
+      closeModal();
+    }
+  };
```
**Можно лучше:**

лучше занести функцию в useEffect.

ArslanMustafin commented on src/components/order-details/order-details.jsx:
```diff
@@ -0,0 +1,21 @@
+function OrderDetails(props) {
+  return (
+    <div className={styles.details}>
+      <p className={styles.text}>{props.orderNumber.toString().padStart(6, '0')}</p>
```
**Можно лучше:**

Лучше вынести в переменную, а то при каждом ререндере будут функции срабатывать))

```js
function OrderDetails({ orderNumber }) {
  const order = orderNumber.toString().padStart(6, '0');

  return (
    // ...
  )
}
```

ArslanMustafin commented on src/utils/burger-api.js:
```diff
@@ -0,0 +1,29 @@
+  }
+}
+
+export default getIngredientsData;
```
**Можно лучше:**

тут будет ещё много функций, поэтому лучше делать просто `export`.

ArslanMustafin reviewed (changes requested):
## Code Review

Здравствуйте!

Отличная работа. 🔥

Приложение начинает обрастать функционалом и уже постепенно проглядывается картина готового продукта.

### **Вы молодец!**

- Декомпозиция файлов по директориям соответствует описанию;
- Сборка и запуск проекта выполняются без ошибок;
- Есть модальные окна;
- Данные получают с сервера;

### **Но есть несколько замечаний, которые нужно исправить:**

- Необходимо указывать .isRequired как у элемента массива, так и у самого массива в propTypes.

  Пример как правильно:

  ```js
  export const dataPropType = PropTypes.arrayOf(ingredientPropType.isRequired);
  ```

- Не стоит писать стили инлайново, в будущем это может затруднить поиск ошибок, и перебить стили заданных классов. Лучше дать какой-то класс, и там указать уже нужные вам стили.

- Также отметил замечания в коде.

### **Сделает код лучше:**

- Рекомендую поправить верстку, сделать её более резиновой, убрать контейнеров фиксированную ширину и высоту. <https://disk.yandex.com/i/pGtDhgqYguwW7A> И выровнять контент по центру. А то на широких экранах контент получается слева.

- Перед отправкой работы убедитесь в том, что весь неиспользуемый код удален.

- Стоит обращать внимание на предупреждения в терминале, в котором запущен проект. Это хорошие рекомендации, которых следовало бы придерживаться. В противном случае, лучше отключить линтер у строки (куска кода) eslint-disable. Так же добавить комментарий (TODO), объясняющий причину блокировки линтера в данном месте. Суть в том, что мы стараемся писать более чистый код, а если видим, что порой выскакивают предупреждения того же линтера, то оставляем комментарий, объясняющий причину. В коммерческой разработке, когда есть целая команда разработчиков, работающих над одним проектом, эта рекомендация послужит правилом хорошего тона. Неплохая статься по поводу локального отключения линтера: <https://learn.coderslang.com/ru/0023-eslint-disable-for-specific-lines-files-and-folders>

### **_Подробно эти и более мелкие замечания, отмечены в коде._**

#### **Хочу напомнить, что работа может быть принята, только после исправления всех "Надо исправить". Перед отправкой на ревью, проверьте работоспособность проекта и наличие возможных ошибок в консоли браузера (кнопка F12).**

dmitry-razumov committed:
refact: isActive to app-header
refact: import { useState, useEffect } from "react"
refact: rm data[] in data.js
refact: get data in app & from api
fix: product[] inside out component
fix: ESlint
refact: product -> products
fix: destruct props ingredientsDetails
fix: forEach Main
fix: orderNumber OrderDetails
fix: export getIngredientsData
refact: closeByEsc
refact: component ShowItem
refact: show tabs
feat: component ShowItem
refactor: BurgerConstructor
fix: ConstructorElement & etc
fix: items (bun & etc) useMemo
add: adaptive to page

dmitry-razumov commented:
for 2nd review
refact: isActive to app-header
refact: import { useState, useEffect } from "react"
refact: rm data[] in data.js
refact: get data in app & from api
fix: product[] inside out component
fix: ESlint
refact: product -> products
fix: destruct props ingredientsDetails
fix: forEach Main
fix: orderNumber OrderDetails
fix: export getIngredientsData
refact: closeByEsc
refact: component ShowItem
refact: show tabs
feat: component ShowItem
refactor: BurgerConstructor
fix: ConstructorElement & etc
fix: items (bun & etc) useMemo
add: adaptive to page

ArslanMustafin reviewed (approved):
## Code Review

Здравствуйте!

Отличная работа.

### **Вы молодец! Работа принята, удачи в дальнейшем обучении!**

- Практически все замечания были исправлены;

### **Но осталось 1 небольшое замечание:**

- Желательно исправить следующие замечания в коде;

  <https://github.com/dmitry-razumov/react-stellar-burger/pull/1#discussion_r1363694275>

### **Сделает код лучше:**

- В некоторых местах нарушено форматирование кода, проблемы с отступами и в некоторых местах слишком длинные строки, что затрудняет чтение кода. Об оформлении кода можно почитать здесь <https://learn.javascript.ru/coding-style>. Постарайтесь использовать автоматическое форматирование, это сильно экономит время. Одно из наиболее популярных дополнений для форматирования кода - Prettier (<https://prettier.io/>)

- Также заметил несколько слов написанных с ошибками. Чтобы это происходило реже попробуйте попользоваться расширением spell checker. <https://marketplace.visualstudio.com/items?itemName=streetsidesoftware.code-spell-checker>
